### PR TITLE
Feature/Meals

### PR DESCRIPTION
**Pull request by rishabh-vinsol**

### CODE REVIEW

File: app/models/ingredient_meal.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  belongs_to :ingredient
+  belongs_to :meal
+
+  ### VALIDATIONS ###
+
+  validates :ingredient_quantity, presence: true
+
+  ### CALLBACKS ###
+
+  after_commit :set_meal_price
```
Comment (himanshumishra31) [inline app/models/ingredient_meal.rb:15]: this should be on create

File: db/migrate/20230522121955_create_join_table_ingredient_meal.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class CreateJoinTableIngredientMeal < ActiveRecord::Migration[7.0]
+  def change
+    create_table :ingredients_meals do |t|
+      t.references :meal
```
Comment (himanshumishra31) [inline db/migrate/20230522121955_create_join_table_ingredient_meal.rb:4]: add meal_id and ingredient unique index

File: db/migrate/20230522121955_create_join_table_ingredient_meal.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+      t.references :ingredient
+      t.integer :ingredient_quantity
```
Comment (himanshumishra31) [inline db/migrate/20230522121955_create_join_table_ingredient_meal.rb:6]: always add default value

File: db/migrate/20230522121423_create_meals.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class CreateMeals < ActiveRecord::Migration[7.0]
+  def change
+    create_table :meals do |t|
+      t.string :name
+      t.integer :price
+      t.boolean :active
+      t.boolean :non_veg
```
Comment (himanshumishra31) [inline db/migrate/20230522121423_create_meals.rb:7]: default values in boolean and integer fields

File: app/models/meal.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+class Meal < ApplicationRecord
+  ### ASSOCIATIONS ###
+
+  has_one_attached :image
+  has_many :ingredients_meals, class_name: 'IngredientMeal'
```
Comment (himanshumishra31) [inline app/models/meal.rb:5]: add dependent conditions in associations everywhere applicable

File: app/models/meal.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  has_many :ingredients, through: :ingredients_meals
+  accepts_nested_attributes_for :ingredients_meals, update_only: true, allow_destroy: true
+
+  ### VALIDATIONS ###
+
+  validates :name, presence: true
+  validates :image, content_type: { in: %w[image/jpeg image/jpg image/png] }
```
Comment (himanshumishra31) [inline app/models/meal.rb:12]: add price validation value greater than 0

File: app/models/meal.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  validates :image, size: { less_than: 5.megabytes }
+
+  def set_non_veg
+    update_column(:non_veg, ingredients.exists?(non_veg: true))
+  end
+
+  def set_price
+    update_column(:price, ingredients_meals.sum { |im| im.ingredient.price_per_portion * im.ingredient_quantity })
```
Comment (himanshumishra31) [inline app/models/meal.rb:20]: add a method in ingredient meal for this logic `im.ingredient.price_per_portion * im.ingredient_quantity`

File: app/models/ingredient_meal.rb (hunk 1 above)
Comment (rishabh-vinsol) [reply app/models/ingredient_meal.rb:15]: Done

File: app/models/meal.rb (hunk 6 above)
Comment (rishabh-vinsol) [reply app/models/meal.rb:12]: Done

File: app/models/meal.rb (hunk 7 above)
Comment (rishabh-vinsol) [reply app/models/meal.rb:20]: Done

File: app/models/meal.rb (hunk 5 above)
Comment (rishabh-vinsol) [reply app/models/meal.rb:5]: Done

File: db/migrate/20230522121955_create_join_table_ingredient_meal.rb (hunk 2 above)
Comment (rishabh-vinsol) [reply db/migrate/20230522121955_create_join_table_ingredient_meal.rb:4]: Done

File: db/schema.rb [hunk 8]
```diff
@@ -84,6 +95,43 @@
...
+    t.string "name"
+    t.integer "price", default: 0
+    t.boolean "active"
+    t.boolean "non_veg", default: false
+    t.text "description"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+  end
+
+  create_table "orders", force: :cascade do |t|
```
Comment (Sparsh-Goyal-2000) [inline db/schema.rb:108]: Why are these changes present in this PR?

File: db/schema.rb [hunk 9]
```diff
@@ -84,6 +95,43 @@
...
+    t.datetime "placed_on"
+    t.datetime "pickup_time"
+    t.datetime "picked_up_at"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["branch_id"], name: "index_orders_on_branch_id"
+    t.index ["user_id"], name: "index_orders_on_user_id"
+  end
+
+  create_table "payments", force: :cascade do |t|
```
Comment (Sparsh-Goyal-2000) [inline db/schema.rb:123]: same

File: db/migrate/20230522121423_create_meals.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class CreateMeals < ActiveRecord::Migration[7.0]
+  def change
+    create_table :meals do |t|
+      t.string :name
+      t.integer :price, default: 0
+      t.boolean :active, deafult: false
```
Comment (Sparsh-Goyal-2000) [inline db/migrate/20230522121423_create_meals.rb:6]: how is this `default` working ?

File: db/schema.rb [hunk 11]
```diff
@@ -95,10 +143,17 @@
     t.string "auth_token"
     t.string "reset_token"
     t.datetime "verified_at", precision: nil
+    t.bigint "branch_id"
```
Comment (Sparsh-Goyal-2000) [inline db/schema.rb:146]: same

File: app/views/meals/_meal.html.erb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  <td><%= meal.id %></td>
+  <td><%= image_tag meal.image, height: 50 if meal.image.attached? %></td>
+  <td><%= meal.name %></td>
+  <td><%= meal.price %></td>
+  <td><%= meal.non_veg ? "Non-Veg" : "Veg" %></td>
+  <td><%= truncate(meal.description, length: 30)%></td>
+  <td>
+    <div class= "d-flex">
+      <% if meal.active %>
+        <%= link_to toggle_inactive_meal_path(meal), class: 'btn btn-success btn-sm me-2', role: 'button' do %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/meals/_meal.html.erb:11]: icon should be reversed
for activating there should be an open eye and vice versa

Comment (Sparsh-Goyal-2000) [inline app/views/meals/_meal.html.erb:11]: Also can we add tooltip saying activate/deactivate on the icon?

File: app/models/meal.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+  has_many :ingredients_meals, class_name: 'IngredientMeal', dependent: :destroy
+  has_many :ingredients, through: :ingredients_meals
+  accepts_nested_attributes_for :ingredients_meals, update_only: true, allow_destroy: true
+
+  ### VALIDATIONS ###
+
+  validates :name, presence: true
+  validates :image, content_type: { in: %w[image/jpeg image/jpg image/png] }
+  validates :image, size: { less_than: 5.megabytes }
+  validates :price, numericality: { greater_than_or_equal_to: 0 }
```
Comment (Sparsh-Goyal-2000) [inline app/models/meal.rb:14]: Getting this error for new meals
image

File: app/controllers/meals_controller.rb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    redirect_to meals_path
+  end
+
+  private def set_meal
+    @meal = Meal.find_by(id: params[:id])
+    redirect_to meals_path, alert: t('errors.meals.not_found') unless @meal
+  end
+
+  private def meals_params
+    params.require(:meal).permit(:name, :price, :active, :non_veg, :description, :image, ingredients_meals_attributes: [:ingredient_id, :ingredient_quantity, :_destroy, :id])
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/meals_controller.rb:52]: are you setting up price in the form?

**Review by Sparsh-Goyal-2000 — CHANGES_REQUESTED** [review]

File: app/controllers/meals_controller.rb (hunk 14 above)
Comment (rishabh-vinsol) [reply app/controllers/meals_controller.rb:52]: Fixed

File: app/views/meals/_meal.html.erb (hunk 12 above)
Comment (rishabh-vinsol) [reply app/views/meals/_meal.html.erb:11]: Done

Comment (rishabh-vinsol) [reply app/views/meals/_meal.html.erb:11]: Done

File: app/controllers/meals_controller.rb (hunk 14 above)
Comment (rishabh-vinsol) [reply app/controllers/meals_controller.rb:52]: Fixed

File: db/schema.rb (hunk 9 above)
Comment (rishabh-vinsol) [reply db/schema.rb:123]: Done

File: db/schema.rb (hunk 8 above)
Comment (rishabh-vinsol) [reply db/schema.rb:108]: Removed, pushed it by mistake

File: app/models/meal.rb (hunk 13 above)
Comment (rishabh-vinsol) [reply app/models/meal.rb:14]: Sir its working fine, please check if it persists in the next PR

File: app/views/meals/_meal.html.erb (hunk 12 above)
Comment (Sparsh-Goyal-2000) [reply app/views/meals/_meal.html.erb]: @rishabh-vinsol 
tooltip not working

File: app/views/meals/_meal.html.erb [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  <td><%= meal.id %></td>
+  <td><%= image_tag meal.image, height: 50 if meal.image.attached? %></td>
+  <td><%= meal.name %></td>
+  <td><%= meal.price %></td>
+  <td><%= meal.non_veg ? "Non-Veg" : "Veg" %></td>
+  <td><%= truncate(meal.description, length: 30)%></td>
+  <td>
+    <div class= "d-flex">
+      <% if meal.active %>
+        <%= link_to toggle_inactive_meal_path(meal), class: 'btn btn-danger btn-sm me-2', role: 'button', title: 'Deactivate' do %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/meals/_meal.html.erb:11]: check tooltip not working

**Review by Sparsh-Goyal-2000 — APPROVED** [review]

**Pull request merged**